bokeh/charts/_builder.py

```diff
@@ -0,0 +1,278 @@
+
+        Args:
+            legend (str, bool): the legend of your plot. The legend content is
+                inferred from incoming input.It can be ``top_left``,
+                ``top_right``, ``bottom_left``, ``bottom_right``.
+                It is ``top_right`` is you set it as True.
+            palette(list, optional): a list containing the colormap as hex values.
+
+        """
+        self._legend = legend
+        self._palette = palette or DEFAULT_PALETTE
+        self._legends = []
+        self.data = {}
+        self.groups = []
+        self.attr = []
+        self.groups = []
+
+    def prepare_values(self):
+        """Prepare the input data.
+
+        Converts data input (self.values) to a DataAdapter and creates
+        instance index if needed
+        """
+        if hasattr(self, 'index'):
+            self.values_index, self.values = DataAdapter.get_index_and_data(
+                self.values, self.index
+            )
+        else:
+            if not isinstance(self.values, DataAdapter):
+                self.values = DataAdapter(self.values, force_alias=False)
+
+    def get_data(self):
+        """Get the input data.
+
+        It has to be implemented by any of the inherited class
+        representing each different chart type. It is the place
+        where we make specific calculations for each chart.
+        """
+        pass
+
+    def get_source(self):
+        """Push data into the ColumnDataSource and build the proper ranges.
+
+        It has to be implemented by any of the inherited class
+        representing each different chart type.
+        """
+        pass
+
+    def draw(self):
+        """Draw the glyphs into the plot.
+
+        It has to be implemented by any of the inherited class
+        representing each different chart type.
+        """
+        pass
+
+    def _pre_create(self):
+        """ Hook method that can be overwritten to inject chart logic to
+        be executed before make renderers method"""
+
+    def make_renderers(self):
+        """
+        Executes chart show core operations:
+
+         - checks for chain methods
+         - prepare chart data & source
+         - create indexes
+         - create glyphs
+         - draw glyphs
+        """
+        # prepare values to for the specific chart type
+        self.prepare_values()
+        # we get the data from the incoming input
+        self.get_data()
+        # we filled the source and ranges with the calculated data
+        self.get_source()
+        # we add the glyphs into the plot
+        return self.draw()
+
+
+    def create(self, chart=None):
+        self._pre_create()
+
+        # pass these renderers to the chart and then forget about the chart
+        renderers = self.make_renderers()
+        chart.add_renderers(self, renderers)
+
+        # create chart ranges..
+        if not chart.x_range:
+            chart.x_range = self.x_range
```

Comment:
All builders need to define a `x_range` and `y_range` attributes. I'm not really happy about it really but the moment I cannot think about a better way of doing this..That's why I mentioned to take a better look at this part...